src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java

```diff
@@ -8,26 +8,108 @@
+                            throw new InvalidMessageException(MESSAGE_TOO_LONG, message);
+                        }
+                        if (sendBuffer.remaining() < (data.length + 1)) {
+                            outboundQueue.put(sendBuffer);
+                            sendBuffer = bufferPool.borrow();
+                        }
+                        if (sendBuffer.position() > 0) {
+                            sendBuffer.put((byte) '\n');
+                        }
+                        sendBuffer.put(data);
+                        if (null == processorWorkQueue[this.processorQueueId].peek()) {
+                            outboundQueue.put(sendBuffer);
+                            sendBuffer = bufferPool.borrow();
+                        }
+                    }
+                } catch (final InterruptedException e) {
+                    if (shutdown) {
+                        endSignal.countDown();
+                        return;
+                    }
+                } catch (final Exception e) {
+                    handler.handle(e);
+                }
+            }
+            endSignal.countDown();
+        }
+    }
 
     StatsDNonBlockingProcessor(final int queueSize, final StatsDClientErrorHandler handler,
-            final int maxPacketSizeBytes, final int poolSize, final int workers)
-            throws Exception {
+            final int maxPacketSizeBytes, final int poolSize, final int workers,
+            final int lockShardGrain) throws Exception {
 
-        super(queueSize, handler, maxPacketSizeBytes, poolSize, workers);
-        this.qsize = new AtomicInteger(0);
+        super(queueSize, handler, maxPacketSizeBytes, poolSize, workers, lockShardGrain);
+        this.qsize = new AtomicInteger[lockShardGrain];
         this.qcapacity = queueSize;
-        this.messages = new ConcurrentLinkedQueue<String>();
+
+        this.messages = new ConcurrentLinkedQueue[lockShardGrain];
+        for (int i = 0 ; i < lockShardGrain ; i++) {
+            this.qsize[i] = new AtomicInteger();
+            this.messages[i] = new ConcurrentLinkedQueue<String>();
+            this.qsize[i].set(0);
+        }
+
+        this.processorWorkQueue = new ConcurrentLinkedQueue[workers];
+        for (int i = 0 ; i < workers ; i++) {
+            this.processorWorkQueue[i] = new ConcurrentLinkedQueue<Integer>();
+        }
     }
 
     @Override
     boolean send(final String message) {
         if (!shutdown) {
-            if (qsize.get() < qcapacity) {
-                messages.offer(message);
-                qsize.incrementAndGet();
+            int threadId = getThreadId();
+            int shard = threadId % lockShardGrain;
+            int processQueue = threadId % workers;
+
+            if (qsize[shard].get() < qcapacity) {
+                messages[shard].offer(message);
+                qsize[shard].incrementAndGet();
+                processorWorkQueue[processQueue].offer(shard);
```

Comment:
Here if I'm correct we are dispatching message to a sharded queue, and those shards are unrelated to processQueue. Did we try 1:1 (or 1:N) mapping between process queues and message queues ? IMHO (unsure) it should (a little bit) ease the concurrency within the JVM.